contracts/ScopeGuard.sol

```diff
@@ -190,32 +210,46 @@ contract ScopeGuard is FactoryFriendly, BaseGuard {
         bytes memory,
         address
     ) external view override {
-        require(
-            operation != Enum.Operation.DelegateCall ||
-                allowedTargets[to].delegateCallAllowed,
-            "Delegate call not allowed to this address"
-        );
-        require(allowedTargets[to].allowed, "Target address is not allowed");
-        if (value > 0) {
-            require(
-                allowedTargets[to].valueAllowed,
-                "Cannot send ETH to this target"
-            );
+        if (data.length != 0 && data.length < 4) {
+            revert FunctionSignatureTooShort();
         }
-        if (data.length >= 4) {
-            require(
-                !allowedTargets[to].scoped ||
-                    allowedTargets[to].allowedFunctions[bytes4(data)],
-                "Target function is not allowed"
-            );
-        } else {
-            require(data.length == 0, "Function signature too short");
-            require(
-                !allowedTargets[to].scoped ||
-                    allowedTargets[to].fallbackAllowed,
-                "Fallback not allowed for this address"
-            );
+
+        Target storage target = targets[to];
+
+        if (
+            value > 0 &&
+            target.options != ExecutionOptions.Send &&
+            target.options != ExecutionOptions.Both
+        ) {
+            // isSend && !canSend
+            revert SendNotAllowed();
         }
+
+        if (
+            operation == Enum.Operation.DelegateCall &&
+            target.options != ExecutionOptions.DelegateCall &&
+            target.options != ExecutionOptions.Both
+        ) {
+            // isDelegateCall && !canDelegateCall
+            revert DelegateCallNotAllowed();
+        }
+
+        if (target.clearance == Clearance.None) {
```

Comment:
We should do this check earlier in the function to save gas on this revert. The other checks are redundant if this one fails.

Reply:
OK